File: ros2/src/shared/march_utility/march_utility/utilities/node_utils.py
# Copyright (C) 2021 Bas Volkers
#
# This program is free software: you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# Version 3 as published by the Free Software Foundation WITH
# additional terms published by Project MARCH per section 7 of
# the GNU General Public License Version 3.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License INCLUDING the additional terms for
# more details.
#
# You should have received a copy of the GNU General Public License
# AND the additional terms along with this program. If not,
# see <https://projectmarch.nl/s/LICENSE> and
# <https://projectmarch.nl/s/LICENSE-ADDITIONAL-TERMS>.

"""This module contains some generic functions for python nodes."""

from typing import Optional, List

import rclpy
from march_shared_msgs.srv import GetJointNames
from rcl_interfaces.srv import GetParameters
from rclpy.node import Node
from rclpy.client import Client
from urdf_parser_py import urdf

SERVICE_TIMEOUT = 1


def get_robot_urdf(node: Node) -> urdf.Robot:
    """Get the robot description from the robot state publisher.

    :param node Node to use for making the request.
    """
    robot_description_client = node.create_client(
        srv_type=GetParameters,
        srv_name="/march/robot_state_publisher/get_parameters",
    )
    wait_for_service(node, robot_description_client, 2)

    robot_future = robot_description_client.call_async(
        request=GetParameters.Request(names=["robot_description"])
    )
    rclpy.spin_until_future_complete(node, robot_future)

    return urdf.Robot.from_xml_string(robot_future.result().values[0].string_value)


def wait_for_service(
    node: Node, client: Client, timeout: Optional[float] = SERVICE_TIMEOUT
):
    """
    Wait for a service to become available.

    :param node: Node to use for logging
    :param client: Client of the service.
    :param timeout: Optional timeout to wait before logging again
    """
    while not client.wait_for_service(timeout_sec=timeout):
        node.get_logger().info(f"Waiting for {client.srv_name} to become available")


def get_joint_names(node: Node) -> List[str]:
    """Get the joint names from the robot information node."""
    joint_names_client = node.create_client(
        srv_type=GetJointNames,
        srv_name="/march/robot_information/get_joint_names",
    )
    wait_for_service(node, joint_names_client)

    future = joint_names_client.call_async(request=GetJointNames.Request())
    rclpy.spin_until_future_complete(node, future)

    return future.result().joint_names

File: ros2/src/shared/march_utility/src/node_utils.cpp
/*
 * Copyright (C) 2021 Bas Volkers, Thijs Raymakers
 *
 * This program is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * Version 3 as published by the Free Software Foundation WITH
 * additional terms published by Project MARCH per section 7 of
 * the GNU General Public License Version 3.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License INCLUDING the additional terms for
 * more details.
 *
 * You should have received a copy of the GNU General Public License
 * AND the additional terms along with this program. If not,
 * see <https://projectmarch.nl/s/LICENSE> and
 * <https://projectmarch.nl/s/LICENSE-ADDITIONAL-TERMS>.
 */

// Copyright 2020 Project March.
#include "march_utility/node_utils.hpp"
#include "march_shared_msgs/srv/get_joint_names.hpp"
#include "rclcpp/rclcpp.hpp"

#include <chrono>
#include <string>
#include <vector>

using namespace std::chrono_literals;

namespace node_utils {
/*
 * @brief Get the Joint names from the robot information node.
 * @param node Node to create client for and use logger of.
 * @return Returns the list of joint names.
 */
std::vector<std::string> get_joint_names(rclcpp::Node& node)
{
    std::vector<std::string> names;

    auto client = node.create_client<march_shared_msgs::srv::GetJointNames>(
        "/march/robot_information/get_joint_names");
    auto request
        = std::make_shared<march_shared_msgs::srv::GetJointNames::Request>();

    // Wait until the service is available. Sending a request to an unavailable
    // service will always fail.
    while (!client->wait_for_service(5s)) {
        if (!rclcpp::ok()) {
            RCLCPP_ERROR(node.get_logger(),
                "Interrupted while waiting for get_joint_names service. "
                "Exiting.");
            return names;
        }
        RCLCPP_INFO(node.get_logger(),
            "The get_joint_names service was unavailable, waiting again...");
    }

    // Send the request and push the received names to the names vector.
    auto result = client->async_send_request(request);
    if (rclcpp::spin_until_future_complete(
            node.get_node_base_interface(), result)
        == rclcpp::FutureReturnCode::SUCCESS) {
        names = std::move(result.get()->joint_names);
    } else {
        RCLCPP_ERROR(
            node.get_logger(), "Failed to call get_joint_names service");
    }

    return names;
}
} // namespace node_utils
